tests/integration_tests/tests/agent_tests/test_plugin_update.py

```diff
@@ -17,63 +17,258 @@
+        return f(self, *args, **kwargs)
+
+    return wrapper
+
+
+def setup_for_plugins_update(f):
+    @wraps(f)
+    def wrapper(self, *args, **kwargs):
+        self.blueprint_name_prefix = 'plugins_update'
+        self.base_blueprint_id = 'b{0}'.format(uuid.uuid4())
+        _f = uploads_mock_plugins(f)
+        return _f(self, *args, **kwargs)
+
+    return wrapper
+
+
+def uploads_mock_plugins(f):
+    @wraps(f)
+    def wrapper(self, *args, **kwargs):
+        self.addCleanup(self._clear_managed_plugins)
+        return f(self, *args, **kwargs)
+
+    return wrapper
 
 
 class TestPluginUpdate(AgentTestWithPlugins):
+    versions = ['1.0', '2.0']
+    dsl_resources_path = resource(os.path.join('dsl', 'agent_tests'))
+    blueprint_name_prefix = 'plugin_update_'
+    setup_deployment_ids = None
+
+    def tearDown(self):
+        deps_to_remove = []
+        if self.setup_deployment_ids:
+            deps_to_remove += self.setup_deployment_ids
+        if self.setup_deployment_id:
+            deps_to_remove.append(self.setup_deployment_id)
+        for dep in deps_to_remove:
+            try:
+                self.client.deployments.delete(dep, ignore_live_nodes=True)
+            except CloudifyClientError:
+                pass
+        super(TestPluginUpdate, self).tearDown()
+
+    @uploads_mock_plugins
     def test_plugin_update(self):
         self.setup_deployment_id = 'd{0}'.format(uuid.uuid4())
         self.setup_node_id = 'node'
         self.plugin_name = 'version-aware-plugin'
         self.base_name = 'base'
+        self.base_blueprint_id = 'b{0}'.format(uuid.uuid4())
         self.mod_name = 'mod'
+        self.mod_blueprint_id = 'b{0}'.format(uuid.uuid4())
 
         # Upload V1.0 and V2.0 plugins
         self.upload_mock_plugin(self.plugin_name)
-        self._upload_v_2_0_plugin()
+        self._upload_v_2_plugin()
+
+        self._upload_blueprints_and_deploy_base()
+
+        # Execute base (V 1.0) workflows
+        self._execute_workflows()
+        self._assert_on_values(self.versions[0])
+
+        self._perform_update()
+
+        # Execute mod (V 2.0) workflows
+        self._execute_workflows()
+        self._assert_on_values(self.versions[1])
 
+    @setup_for_sourced_plugins
+    def test_sourced_plugin_updates(self):
         self._upload_blueprints_and_deploy_base()
 
         # Execute base (V 1.0) workflows
         self._execute_workflows()
-        self._assert_on_values('1.0')
+        self._assert_on_values(self.versions[0])
 
         self._perform_update()
 
         # Execute mod (V 2.0) workflows
         self._execute_workflows()
-        self._assert_on_values('2.0')
+        self._assert_on_values(self.versions[1])
 
-    def _perform_update(self):
-        self.client.deployment_updates.update_with_existing_blueprint(
-            deployment_id=self.setup_deployment_id,
-            blueprint_id=self.mod_name
+    @setup_for_sourced_plugins
+    def test_sourced_plugin_doesnt_update(self):
```

Comment:
test_sourced_plugin_doesnt_update -> test_sourced_plugin_doesnt_update_with_same_version

Reply:
Then it might sound as if the blueprint had the same version or something, but here it explicitely didn't upate the plugins although the node data shows it has plugin v2.

Reply:
but there is more then one reason for this to happen

Reply:
I don't quite understand, in here I pass `update_plugins=False` and expect the plugins don't update. Why not use that as a name?

Reply:
because that connection it's clear, try to view it if a person that wants to understand what the feature does without reading the code or a doc

Reply:
`test_sourced_plugin_remains_with_same_version`?